src/Microsoft.Azure.WebJobs.Host/Scale/ConcurrencyManagerService.cs

```diff
@@ -127,10 +127,11 @@ internal async Task WriteSnapshotAsync()
         {
             var snapshot = _concurrencyManager.GetSnapshot();
 
-            // only persist snapshots for functions that are in our current index
-            // this ensures we prune any stale entries from a previously applied snapshot
+            // Only persist snapshots for functions that are in our current index. This ensures
+            // we prune any stale entries from a previously applied snapshot.
+            // Note that functions using a shared listener are treated specially here.
             var functionIndex = await _functionIndexProvider.GetAsync(CancellationToken.None);
-            var functions = functionIndex.ReadAll().ToLookup(p => p.Descriptor.Id, StringComparer.OrdinalIgnoreCase);
+            var functions = functionIndex.ReadAll().ToLookup(p => p.Descriptor.SharedListenerId ?? p.Descriptor.Id, StringComparer.OrdinalIgnoreCase);
```

Comment:
This was the other issue I discovered in testing DC with BlobTrigger end to end. The issue was with this logic that prunes snapshots for IDs that don't exist in the function index. The DC snapshot IDs use the shared listener ID, but the function index previously didn't contain any information on the shared listener ID. That meant we'd never persist snapshots for these functions. Now, for functions using a shared listener, we use the shared ID and everything lines up.